daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala

```diff
@@ -6,43 +6,67 @@ package data
 
 import com.daml.lf.data.Ref.Name
 
-final case class Struct[+X] private (sortedFields: ImmArray[(Ref.Name, X)]) extends NoCopy {
+/** We use this container to describe structural record as sorted flat list in various parts of the codebase.
+    `entries` are sorted by their first component without duplicate.
+  */
+final case class Struct[+X] private (private val sortedFields: ImmArray[(Ref.Name, X)])
+    extends NoCopy {
 
-  def lookup(name: Ref.Name): Option[X] =
-    sortedFields.find(_._1 == name).map(_._2)
+  /** O(n) */
+  @throws[IndexOutOfBoundsException]
+  def apply(name: Ref.Name): X = sortedFields(indexOf(name))._2
 
-  def mapValue[Y](f: X => Y) = new Struct(sortedFields.map { case (k, v) => k -> f(v) })
+  /** O(n) */
+  def indexOf(name: Ref.Name): Int = sortedFields.indexWhere(_._1 == name)
 
+  /** O(n) */
+  def lookup(name: Ref.Name): Option[X] = sortedFields.find(_._1 == name).map(_._2)
+
+  /** O(n) */
+  def mapValues[Y](f: X => Y) = new Struct(sortedFields.map { case (k, v) => k -> f(v) })
+
+  /** O(1) */
   def toImmArray: ImmArray[(Ref.Name, X)] = sortedFields
 
+  /** O(1) */
   def names: Iterator[Ref.Name] = iterator.map(_._1)
 
+  /** O(1) */
   def values: Iterator[X] = iterator.map(_._2)
 
+  /** O(1) */
   def iterator: Iterator[(Ref.Name, X)] = sortedFields.iterator
 
-  def foreach(f: ((Ref.Name, X)) => Unit): Unit = sortedFields.foreach(f)
+  /** O(1) */
+  def size: Int = sortedFields.length
+
+  /** O(n) */
+  override def toString: String = iterator.mkString("Struct(", ",", ")")
 
 }
 
 object Struct {
 
-  def apply[X](fields: (Name, X)*): Struct[X] =
-    new Struct(fields.sortBy(_._1: String).to[ImmArray])
-
-  def apply[X](fields: ImmArray[(Name, X)]): Struct[X] = apply(fields.toSeq: _*)
-
-  def fromSortedImmArray[X](fields: ImmArray[(Ref.Name, X)]): Either[String, Struct[X]] = {
-    val struct = new Struct(fields)
-    Either.cond(
-      (struct.names zip struct.names.drop(1)).forall { case (x, y) => (x compare y) <= 0 },
-      struct,
-      s"the list $fields is not sorted by name"
-    )
+  /** Constructs a Struct.
+    * In case one of the field name is duplicated, return it as Left.
+    * O(n log n)
+    */
+  def fromSeq[X](fields: Seq[(Name, X)]): Either[Name, Struct[X]] = {
+    val sortedFields = fields.sortBy(_._1: String)
+    (sortedFields.iterator zip sortedFields.iterator.drop(1))
+      .collectFirst {
+        case ((name1, _), (name2, _)) if name1 == name2 => name1
+      }
+      .toLeft(Struct(ImmArray(sortedFields)))
```

Comment:
I assume we call this whenever we construct a struct. Thus, I would say this is somewhat performance sensitive. I'm wondering if this code is a bit slower than it needs to be because of all the abstractions? How about something like
```suggestion
    var i = 1
    while (i < sortedFields.lenght) {
      if (sortedFields(i-1)._1 == sortedFields(i)._1) {
        return Left(sortedFields(i)._1)
      }
      i += 1
    }
    Right(Struct(ImmArray(sortedFields)))
```

Reply:
In fact no, we will not call this whenever we construct a Struct.
In general we will try to use `mapValues` that is O(n) instead O(n log n).